### DISCUSSION TITLE
`call-interactively` does not use `consult-customize`

### DISCUSSION THREAD

**Discussion by ErnestKz** [minad/consult#1023] [Q&A]
For example, when customizing `consult-find`:
  ```elisp
  (consult-customize
   consult-find
   :prompt "hello")
```

Invoking it via extended command yields a prompt of "hello".

Whereas `(call-interactively 'consult-find)` uses the defaults, ignoring the customized version.

**Comment by ErnestKz** [comment]
Well, I now realise it's not a bug, but a little bit of a tragedy :(

`this-command` is not set by `call-interactively`, which `consult--customize-get` requires to function.
```elisp
(defun consult--customize-get (&optional cmd)
  "Get configuration from `consult--customize-alist' for CMD."
  (mapcar (lambda (x) (eval x 'lexical))
          (alist-get (or cmd this-command) consult--customize-alist)))
```

**Reply by minad** [reply]
What do you want to achieve?

**Reply by ErnestKz** [reply]
I think it would makes sense for `consult-customize` to work for commands called with `call-interactively`. 

In my particular use case, I use the consult commands as a part of `project-switch-project`:
```elisp
(add-to-list 'project-switch-commands '(consult-grep "Grep"))
(add-to-list 'project-switch-commands '(consult-find "Find"))
```
Which in turn runs the commands via `call-interactively`. It was to my suprise that my `consult-customize`ations weren't working when invoked this way.

**Reply by ErnestKz** [reply]
My hopefully possibly temporary workaround is to add advice to the consult functions of interest to always set the `this-command` variable.

```elisp
;; -*- lexical-binding: t -*-
(defun consult-add-this-command-advice (symbol)
  (advice-add
   symbol :around
   (lambda (orig-fn &rest args)
     (let ((real-this-command this-command))
       (setq this-command symbol)
       (let ((res (apply orig-fn args)))
	 (setq this-command real-this-command)
	 res))
     )))
```
```elisp
  (consult-add-this-command-advice 'consult-theme)
  (consult-add-this-command-advice 'consult-buffer)
  (consult-customize
   consult-theme
   consult-buffer
   :preview-key "M-.")

  (consult-add-this-command-advice 'consult-find)
  (consult-customize
   consult-find
   :state (consult--file-preview)
   :preview-key "M-.")

  (consult-add-this-command-advice 'consult-line)
  (consult-customize
   consult-line
   :prompt "Search: "
   :preview-key "M-.")
```

**Reply by ErnestKz** [reply]
I just discovered that `vertico-multiform-comands` also relies on the consult to have `this-command` set to itself. The workaround even solves this, but unfortunately would make a comprehensive fix a bit more involved.

**Reply by minad** [reply]
No, sorry. This should be fixed in `project-switch-commands`. Maybe you can also add an advice there.

**Reply by ErnestKz** [reply]
I'm happy with my workaround, but are you saying `project-switch-commands` should use a modified `call-interactively` such that it sets the `this-command`? I think it make sense, since `project-switch-commands` for a good few cases calls commands that are intended to be used actually interactively.

Do you think such an ammendment to `project-switch-commands` would make even more sense if emacs shipped with such a native function?

**Reply by ErnestKz** [reply]
If you may be graceful enough to receive my contrary perspective against `consult`s and `vertico`s usage of `this-command`, perhaps `previous-command` and `this-command` is intended and understood to be used in-conjunction such that the elisp can know if the user is chaining inputted commands via key-presses. 

Rather how we use it here for `consult` and `vertico` as a variable to store things for a command, because there are other mechanisms which can offer that, and also work in more justified use cases. (i.e defining a new variable with its own name just for use to look-up, in the calling command)

**Reply by minad** [reply]
Yes, using `this-command` for looking up configuration is brittle and may break in special cases, where commands are chained. But we don't have better mechanisms unfortunately. For `vertico-multiform` we can also use the completion category, but of course this works only if the command defines a category in the first place.

**Reply by minad** [reply]
> I'm happy with my workaround, but are you saying project-switch-commands should use a modified call-interactively such that it sets the this-command? I think it make sense, since project-switch-commands for a good few cases calls commands that are intended to be used actually interactively.

Not a modified version of `call-interactively`. It could simply set `this-command` before.

**Reply by ErnestKz** [reply]
It seems you're perceiving a slightly jumbled up meaning of what I'm saying. I'm saying `this-command` is mostly intended to be used with `last-command` in order to detect a chain of commands key'd in by the user. Like the first example on this [page](https://www.gnu.org/software/emacs/manual/html_node/eintr/last_002dcommand-_0026-this_002dcommand.html)

> we don't have better mechanisms unfortunately

Surely there are other mechanisms, maybe not as syntactically elegant, and requiring some extra code.

> Not a modified version of call-interactively. It could simply set this-command before.

Mmm right, and it would have to set `previous-command` to the `this-command` before that (perhaps? i havent though it too much through)

**Reply by minad** [reply]
Maybe I've misunderstood you. However this-command and last-command can be used for many different things, e.g., redirecting commands in a pre-command-hook or detecting the current command as we do here. In any case, the mechanism will stay as is, as long as no better alternative comes along. If you have a better idea, please let me know. But to be clear, I am likely not going to accept it, if it leads to much more code, or has other downsides. Also note that the current mechanism has advantages, for example if you define an alias of a command, it can get its own configuration. This is intentional and wouldn't be possible if we hard-code the command name, for example passing the current command symbol to `consult--read` (either implicitly or via some macro or stack trace trickery).
